### PR TITLE
fix negative samples in nt_xent loss

### LINKED ISSUE #252: Little Bug in calculation of negative samples in nt_xent loss
## 🐛 Bug

I think there is a error in the calculation of the [nt_xent loss](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/c60e142905cf44086d5c3e1ed844953730668f98/pl_bolts/losses/self_supervised_learning.py#L8) for self-supervised learning . 

When calculating the indices of negative samples:

`mask = ~torch.eye(n_samples, device=sim.device).bool()`

the indices of positive samples are calculated and then negated. However only the diagonal is considered here, which contains the Self-correlations, thus simply only ones. The true positive samples are located in the right upper and left lower diagonal. That means, simply said, instead of 

loss = -log(pos/neg), 

loss=-log(pos/(pos+neg))

 is calculated. 

In other words, the tensor "neg" should contain only 2N * 2(N-1) samples, because there are exactly 2(N-1) negative samples and one positive sample per data point. In the current code the tensor contains 2N * (2N-1) samples.

I will submit a pull request referring to this issue with an easy solution.

Best, 

Temesgen

### PR DESCRIPTION
**Pull request by Alokaado**
## What does this PR do?
Fixes #252.

### CODE REVIEW

File: pl_bolts/losses/self_supervised_learning.py [hunk 1]
```diff
@@ -17,7 +17,10 @@ def nt_xent_loss(out_1, out_2, temperature):
     sim = torch.exp(cov / temperature)
 
     # Negative similarity
-    mask = ~torch.eye(n_samples, device=sim.device).bool()
+    diag = torch.eye(n_samples, device=sim.device)
```
Comment (Borda) [inline pl_bolts/losses/self_supervised_learning.py:20]: can we add a test for this edit...

**Comment by williamFalcon** [comment]
@ananyahjha93

**Comment by ananyahjha93** [comment]
The denominator does contain the positive sample itself, so the calculation checks out to be fine. We do not work with log(pos/neg), but rather log(pos / (pos + sum over neg)). This is not a bug, but rather how the contrastive loss is designed.
